pest/src/inputs_private/string_input.rs

```diff
@@ -0,0 +1,305 @@
+// pest. The Elegant Parser
+// Copyright (C) 2017  Dragoș Tiselice
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::ascii::AsciiExt;
+use std::ops::Range;
+use std::str;
+
+use super::Input;
+
+/// A `struct` useful for matching heap-allocated `String`s.
+#[derive(Debug)]
+pub struct StringInput {
+    string: String
+}
+
+impl StringInput {
+    /// Creates a new `StringInput` from a `String`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest::inputs::{Input, StringInput};
+    /// let input = StringInput::new("asd".to_owned());
+    ///
+    /// assert_eq!(input.len(), 3);
+    /// ```
+    pub fn new(string: String) -> StringInput {
+        StringInput {
+            string: string
+        }
+    }
+}
+
+impl Input for StringInput {
+    #[inline]
+    fn len(&self) -> usize {
+        self.string.len()
+    }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        self.string.is_empty()
+    }
+
+    #[inline]
+    unsafe fn slice(&self, start: usize, end: usize) -> &str {
+        self.string.slice_unchecked(start, end)
+    }
+
+    unsafe fn line_col(&self, pos: usize) -> (usize, usize) {
+        if pos > self.string.len() {
+            panic!("position out of bounds");
+        }
+
+        let mut pos = pos;
+        let slice = &self.string[..pos];
+        let mut chars = slice.chars().peekable();
+
+        let mut line_col = (1, 1);
+
+        while pos != 0 {
+            match chars.next() {
+                Some('\r') => {
+                    if let Some(&'\n') = chars.peek() {
+                        chars.next();
+
+                        if pos == 1 {
+                            pos -= 1;
+                            line_col = (line_col.0 + 1, 1);
+                        } else {
+                            pos -= 2;
+                            line_col = (line_col.0 + 1, 1);
```

Comment:
The whole method is a bit ugly. I've invested some time into rethinking it, but gave up. I'm open to other designs here.